File: blade-ai/tui/src/utils/result.ts
/**
 * Parse a Python ``result`` event's content (a JSONEnvelope-shaped
 * string) into a ResultItem suitable for ResultCard rendering.
 *
 * Envelope shape (best-effort):
 *   {
 *     "status": "success" | "fail" | ...,
 *     "data": {
 *       "task_id": "...",
 *       "fault_type": "...",
 *       "blade_uid": "...",
 *       "duration_ms": 95400,
 *       "verification": "...",
 *       "side_effects": { "container_restarts": [...] },
 *       "failure_reason": "<base> | llm_analysis: <hint>",
 *       ...
 *     }
 *   }
 *
 * We tolerate missing fields and never throw — anything we can't read
 * collapses to an empty string in the ResultItem, and the card hides
 * empty rows.
 */

import type { ResultItem } from "../state/types.js";

interface Envelope {
  status?: string;
  data?: Record<string, unknown>;
  message?: string;
}

export function parseResultEnvelope(
  raw: string,
  fallbackTaskId: string,
): Omit<ResultItem, "kind" | "id"> {
  let env: Envelope = {};
  try {
    env = JSON.parse(raw) as Envelope;
    if (typeof env !== "object" || env === null) env = {};
  } catch {
    // Not JSON — return a degenerate result with raw text in summary.
    return {
      taskId: fallbackTaskId,
      status: "unknown",
      faultType: "",
      bladeUid: "",
      duration: "",
      summary: raw.split("\n").find((l) => l.trim()) ?? "",
    };
  }

  const data = (env.data ?? {}) as Record<string, unknown>;
  const taskId = asString(data["task_id"]) || fallbackTaskId;
  const faultType = asString(data["fault_type"]);
  const bladeUid = asString(data["blade_uid"]);
  const duration = formatDurationMs(asNumber(data["duration_ms"]));
  const verification = asString(data["verification"]);

  // Side effects → terse summary chip ("HPA scaled · 1 pod restarted")
  // PLUS a per-row list (e.g. "1 pod restart · accounting-6fb...qn2vr")
  // surfaced in the ResultCard "Side effects" section. The single
  // summary chip is kept for back-compat with the legacy single-row
  // ResultCard layout; the new ``sideEffects`` array drives the v3
  // multi-row section.
  const sideEffectsRaw = data["side_effects"];
  let restartCount = 0;
  const sideEffectList: string[] = [];
  if (typeof sideEffectsRaw === "object" && sideEffectsRaw !== null) {
    const restarts = (sideEffectsRaw as Record<string, unknown>)["container_restarts"];
    if (Array.isArray(restarts)) {
      restartCount = restarts.length;
      for (const r of restarts) {
        if (typeof r === "string") {
          sideEffectList.push(`pod restart · ${r}`);
        } else if (r && typeof r === "object") {
          const obj = r as Record<string, unknown>;
          const name = asString(obj["name"]) || asString(obj["pod"]);
          const reason = asString(obj["reason"]);
          const label = name
            ? reason
              ? `pod restart · ${name} (${reason})`
              : `pod restart · ${name}`
            : "pod restart";
          sideEffectList.push(label);
        }
      }
    }
    // Other side-effect categories the agent may surface in future
    // (HPA scaling, log warnings, …). Flatten anything iterable into
    // a "<key> · <value>" row so additions to the server contract
    // light up here without a client release.
    for (const [k, v] of Object.entries(sideEffectsRaw as Record<string, unknown>)) {
      if (k === "container_restarts") continue;
      if (Array.isArray(v)) {
        for (const item of v) {
          sideEffectList.push(
            `${k} · ${typeof item === "string" ? item : JSON.stringify(item)}`,
          );
        }
      }
    }
  }

  // Replan count — surfaced in Outcome section as "succeeded after N
  // auto-replan(s)". 0 means clean first-try; we omit the field then.
  const replanCount = asNumber(data["replan_count"]);

  // Live target spec — verifies "did we hit the intended target".
  let target: { namespace?: string; names?: string[] } | undefined;
  const rawTarget = data["target"];
  if (rawTarget && typeof rawTarget === "object" && !Array.isArray(rawTarget)) {
    const t = rawTarget as Record<string, unknown>;
    const ns = asString(t["namespace"]);
    const namesRaw = t["names"];
    const names = Array.isArray(namesRaw)
      ? namesRaw.map(asString).filter(Boolean)
      : [];
    if (ns || names.length > 0) {
      target = { namespace: ns || undefined, names: names.length ? names : undefined };
    }
  }

  // Compose a single-line effect summary. Verification often itself
  // is a multi-line string — we take the first non-empty line, strip
  // emoji-style ✓/✗ prefixes the agent loves to add.
  const verifLine =
    verification
      .split("\n")
      .map((l) => l.trim().replace(/^[✓✗•\-•]+\s*/, ""))
      .find((l) => l.length > 0) ?? "";

  const parts: string[] = [];
  if (verifLine) parts.push(verifLine);
  if (restartCount > 0) parts.push(`${restartCount} pod restart${restartCount === 1 ? "" : "s"}`);
  const summary = parts.join(" · ");

  // Failure cause / hint — prefer structured failure_detail over legacy failure_reason.
  let cause: string | undefined;
  let hint: string | undefined;
  const failureDetail = data["failure_detail"] as Record<string, unknown> | undefined;
  if (failureDetail && typeof failureDetail === "object" && failureDetail["category"]) {
    const category = asString(failureDetail["category"]);
    const context = asString(failureDetail["context"]);
    cause = context ? `${category}: ${context}` : category;
    hint = asString(failureDetail["llm_analysis"]) || undefined;
  } else {
    // Legacy fallback: failure_reason shaped "<base> | llm_analysis: <hint>"
    const failureReason = asString(data["failure_reason"]);
    if (failureReason) {
      const sep = " | llm_analysis: ";
      if (failureReason.includes(sep)) {
        const idx = failureReason.indexOf(sep);
        cause = failureReason.slice(0, idx).trim();
        hint = failureReason.slice(idx + sep.length).trim();
      } else {
        cause = failureReason;
      }
    }
  }

  // Status mapping. The Python envelope's outer ``status`` is just
  // "ok" / "fail" of the API call itself; the inject's actual outcome
  // lives in ``data.task_state`` or is implied by the presence of a
  // failure_reason.
  const taskState = asString(data["task_state"]);
  let status: ResultItem["status"] = "unknown";
  if (taskState === "injected" || taskState === "recovered") status = "success";
  else if (taskState === "partial_recovered") status = "partial";
  else if (taskState === "failed" || cause) status = "failed";
  // ``ResponseStatus`` enum from Python is "success" | "fail" — see
  // models/schemas.py. Anything else falls through to "unknown".
  else if (env.status === "success") status = "success";

  // T6 — postmortem payload. Server sends ``null`` when generation
  // was skipped (disabled / non-inject / non-whitelist failure /
  // LLM timeout); we leave the field undefined in that case so
  // PostmortemSection can shortcut on `!item.postmortem`.
  let postmortem: ResultItem["postmortem"];
  const rawPm = data["postmortem"];
  if (rawPm && typeof rawPm === "object" && !Array.isArray(rawPm)) {
    const pm = rawPm as Record<string, unknown>;
    const path = asString(pm["path"]);
    const markdown = asString(pm["markdown"]);
    const pmSummary = asString(pm["summary"]);
    if (path && markdown) {
      postmortem = { path, markdown, summary: pmSummary };
    }
  }

  return {
    taskId,
    status,
    faultType,
    bladeUid,
    duration,
    summary,
    cause,
    hint,
    target,
    replanCount: replanCount > 0 ? replanCount : undefined,
    sideEffects: sideEffectList.length > 0 ? sideEffectList : undefined,
    sideEffectsSummary: asString(data["side_effects_summary"]) || undefined,
    postmortem,
  };
}

function asString(v: unknown): string {
  return typeof v === "string" ? v : "";
}
function asNumber(v: unknown): number {
  return typeof v === "number" && Number.isFinite(v) ? v : 0;
}

function formatDurationMs(ms: number): string {
  if (!ms || ms <= 0) return "";
  if (ms < 1000) return `${ms}ms`;
  if (ms < 60_000) return `${Math.round(ms / 1000)}s`;
  const m = Math.floor(ms / 60_000);
  const s = Math.round((ms % 60_000) / 1000);
  return s > 0 ? `${m}m${s.toString().padStart(2, "0")}s` : `${m}m`;
}

File: blade-ai/src/chaos_agent/tui/renderers/result.py
"""Result renderer — structured result cards for task outcomes.

Renders TaskResult envelopes as visually distinct cards with status-colored
borders, structured metadata, and verification sections.
"""

from __future__ import annotations

from rich.console import Group
from rich.panel import Panel
from rich.text import Text

from chaos_agent.tui.console import ChaosConsole
from chaos_agent.tui.renderers._layout import make_field_table
from chaos_agent.tui.state import DisplayMode
from chaos_agent.tui.theme import Borders, Colors, Icons, Spacing
from chaos_agent.utils.time import parse_iso_timestamp


def _read_diagnostic(data: dict, key: str, default):
    """Read a diagnostic field, tolerating envelope vs flat shapes.

    Stream events arrive as a JSONEnvelope (``{status, data: {...}, ...}``),
    while tests pass a flat dict for simplicity. Try the top level first
    (test path) then fall back to ``envelope["data"]`` (production path).
    Returning the inner value lets callers stay shape-agnostic.
    """
    if key in data:
        value = data[key]
        if value is not None:
            return value
    payload = data.get("data")
    if isinstance(payload, dict):
        return payload.get(key, default)
    return default


def _extract_failure_cause_hint(data: dict) -> tuple[str, str]:
    """Extract (cause, hint) from either structured ``failure_detail`` or legacy ``failure_reason``.

    Prefers structured: ``failure_detail.category`` + ``failure_detail.context`` as cause,
    ``failure_detail.llm_analysis`` as hint.
    Falls back to legacy split on ``" | llm_analysis: "`` separator.
    """
    detail = _read_diagnostic(data, "failure_detail", None)
    if isinstance(detail, dict) and detail.get("category"):
        cause = detail["category"]
        ctx = detail.get("context", "")
        if ctx:
            cause = f"{cause}: {ctx}"
        hint = detail.get("llm_analysis", "") or ""
        return cause, hint

    # Legacy fallback
    reason = _read_diagnostic(data, "failure_reason", "") or ""
    if not reason:
        return "", ""
    sep = " | llm_analysis: "
    if sep in reason:
        base, _, analysis = reason.partition(sep)
        return base.strip(), analysis.strip()
    return reason.strip(), ""


# Cap each replan reason to keep the timeline scannable in narrow terminals.
# Longer strings get truncated with an ellipsis — the full value remains
# in build_status_data for /history and any future expansion view.
_REPLAN_REASON_MAX_LEN = 80


def _truncate_reason(text: str, limit: int = _REPLAN_REASON_MAX_LEN) -> str:
    text = (text or "").strip().splitlines()[0] if text else ""
    if len(text) > limit:
        return text[: limit - 1].rstrip() + "\u2026"
    return text


def _render_side_effects(body: Text, side_effects: dict) -> None:
    """Render all side-effect categories detected by the se_detect node."""
    if not side_effects or not any(
        isinstance(v, list) and v for v in side_effects.values()
    ):
        return

    body.append("\n")
    body.append("  \u2500\u2500 Side Effects ", style=Colors.DIM)
    body.append("\u2500" * 32 + "\n", style=Colors.DIM)

    for key, items in side_effects.items():
        if not isinstance(items, list) or not items:
            continue
        for entry in items:
            if not isinstance(entry, dict):
                continue
            _render_side_effect_entry(body, key, entry)


def _render_side_effect_entry(body: Text, category: str, entry: dict) -> None:
    """Render a single side-effect entry with category-aware formatting."""
    body.append(f"  {Icons.WARNING} ", style=Colors.WARNING)

    if category == "container_restarts":
        pod = entry.get("pod") or "unknown"
        count = entry.get("restart_count") or entry.get("restart_delta")
        reason = entry.get("reason") or entry.get("note") or ""
        body.append(f"{pod} restarted", style=Colors.WARNING)
        if isinstance(count, int) and count > 0:
            body.append(f" {count}\u00d7", style=f"bold {Colors.WARNING}")
        if reason:
            body.append(f" \u2014 {_truncate_reason(reason)}", style=Colors.MUTED)
    else:
        label = category.replace("_", " ").title()
        name = (
            entry.get("pod") or entry.get("service") or
            entry.get("hpa") or entry.get("pattern") or ""
        )
        detail = entry.get("reason") or entry.get("message") or ""
        body.append(f"{label}: {name}", style=Colors.WARNING)
        if detail:
            body.append(f" \u2014 {_truncate_reason(detail)}", style=Colors.MUTED)

    body.append("\n")


def _render_replan_history(
    body: Text,
    history: list,
    final_status: str,
) -> None:
    """Render the agent's self-improvement loop as a versioned timeline.

    Each ``replan_history`` entry records the **previous** plan's failure
    that triggered the next attempt — i.e., entry with ``attempt=N`` means
    "v(N-1) failed, agent generated v(N)". We render it as
    ``v1 ✗ <reason> → v2`` so users can see the convergence story.

    Skipped entirely when there were no replans (single-shot success).
    Why surface it: the value pitch of an agent over a static playbook is
    "it learns mid-flight". When that loop fires we should show it.
    """
    if not history:
        return

    body.append("\n")
    body.append("  \u2500\u2500 Replan History ", style=Colors.DIM)
    body.append("\u2500" * 31 + "\n", style=Colors.DIM)

    for entry in history:
        if not isinstance(entry, dict):
            continue
        attempt = int(entry.get("attempt") or 0)
        if attempt <= 0:
            continue
        # entry.attempt=N means "the Nth replan event was just triggered" —
        # i.e., plan v{N} (1-indexed: v1 = original) is the version that failed
        # and the agent is about to run v{N+1}. We surface the failed version.
        prev_v = attempt
        reason = _truncate_reason(entry.get("original_error") or "(no reason recorded)")
        body.append(f"  v{prev_v} ", style=f"bold {Colors.MUTED}")
        body.append(f"{Icons.FAIL} ", style=Colors.ERROR)
        body.append(f"{reason}\n", style=Colors.MUTED)

    final_v = max(int(e.get("attempt") or 0) for e in history if isinstance(e, dict)) + 1
    body.append(f"  v{final_v} ", style=f"bold {Colors.MUTED}")
    if final_status == "success":
        body.append(f"{Icons.SUCCESS} ", style=Colors.SUCCESS)
        body.append("agent improved the plan and succeeded\n", style=Colors.SUCCESS)
    else:
        body.append(f"{Icons.FAIL} ", style=Colors.ERROR)
        body.append("final attempt also failed\n", style=Colors.ERROR)


def _format_offset(seconds: float) -> str:
    """Render a wall-clock offset as ``T+M:SS`` (or ``T+S.s`` when sub-minute).

    Used by the physical timeline so the eye reads "this many seconds
    after the run started" without doing the timestamp math itself.
    Negative offsets degrade to ``T+0`` rather than printing ``T-…`` —
    a clock skew between the task store and the local box shouldn't
    surface as a confusing sentinel.
    """
    if seconds <= 0:
        return "T+0"
    if seconds < 60:
        return f"T+{seconds:.1f}s"
    minutes = int(seconds // 60)
    secs = int(seconds % 60)
    return f"T+{minutes}:{secs:02d}"


def _build_timeline_events(data: dict) -> list[tuple[str, float, str]]:
    """Pull ``(label, offset_seconds, glyph_color)`` rows from the result envelope.

    All timestamps are already collected by the agent; we only re-arrange
    them. We deliberately don't fabricate replan-event timestamps — the
    `replan_history` rows lack their own ``at`` field and an interpolated
    midpoint would be misleading. Returns an empty list when ``created_at``
    is missing (everything else is anchored against it).
    """
    created_at = data.get("created_at") or _read_diagnostic(data, "created_at", "")
    if not created_at:
        return []
    try:
        anchor = parse_iso_timestamp(created_at)
    except (ValueError, TypeError):
        return []

    events: list[tuple[str, float, str]] = [
        ("\u4efb\u52a1\u542f\u52a8", 0.0, Colors.MUTED),  # 任务启动
    ]

    injection_start = data.get("injection_start_time") or _read_diagnostic(
        data, "injection_start_time", ""
    )
    if injection_start:
        try:
            offset = (parse_iso_timestamp(injection_start) - anchor).total_seconds()
            events.append(("\u6545\u969c\u6ce8\u5165", offset, Colors.WARNING))  # 故障注入
        except (ValueError, TypeError):
            pass

    finished_at = data.get("finished_at") or _read_diagnostic(data, "finished_at", "")
    if finished_at:
        try:
            offset = (parse_iso_timestamp(finished_at) - anchor).total_seconds()
            status = data.get("status", "")
            label = (
                "\u4efb\u52a1\u5b8c\u6210" if status == "success"
                else "\u4efb\u52a1\u7ed3\u675f"
            )  # 任务完成 / 任务结束
            color = Colors.SUCCESS if status == "success" else Colors.ERROR
            events.append((label, offset, color))
        except (ValueError, TypeError):
            pass

    return events


def _render_physical_timeline(
    body: Text, data: dict, display_mode: DisplayMode
) -> None:
    """Render the wall-clock timeline (PR-D5).

    Three branches by density:

    * ``calm``    — silent. The user opted out of the differentiating UI.
    * ``working`` — single line: total duration + a one-line summary
      "故障注入 → 完成". Compact enough to fit beside the status fields.
    * ``dense``   — full ``T+0 / T+inject / T+done`` table so a postmortem
      reader can lift the timeline into a doc without reformatting.

    The data is *already* in the envelope (``created_at``,
    ``injection_start_time``, ``finished_at``, ``duration_ms``); this
    function only recomposes it.
    """
    if display_mode == DisplayMode.CALM:
        return
    events = _build_timeline_events(data)
    if not events:
        return

    body.append("\n")
    body.append("  \u2500\u2500 Timeline ", style=Colors.DIM)
    body.append("\u2500" * 33 + "\n", style=Colors.DIM)

    if display_mode == DisplayMode.DENSE:
        for label, offset, color in events:
            body.append(f"  {_format_offset(offset)} ", style=Colors.DIM)
            body.append(f"{label}\n", style=color)
        return

    # working — collapse to a single line: total duration first, then a
    # short arrow-joined summary so the user sees both "how long" and
    # "what fired" without reading three rows.
    duration_ms = data.get("duration_ms") or _read_diagnostic(data, "duration_ms", 0)
    total_seconds = (duration_ms or 0) / 1000.0
    if total_seconds <= 0 and len(events) >= 2:
        total_seconds = events[-1][1] - events[0][1]

    summary_labels = [label for label, _offset, _color in events]
    summary = " \u2192 ".join(summary_labels)
    body.append(
        f"  {_format_offset(total_seconds)} \u00b7 ",
        style=Colors.DIM,
    )
    body.append(f"{summary}\n", style=Colors.MUTED)


def _latest_experiment_locator(state) -> str:
    """Return the most recently allocated experiment locator (e.g. 'E1'), or ''."""
    if state is None:
        return ""
    locators = getattr(state, "locators", None)
    if locators is None:
        return ""
    experiments = locators.list_experiments()
    if not experiments:
        return ""
    return experiments[-1].locator


def render_result(
    console: ChaosConsole,
    data: dict,
    task_id: str = "",
    display_mode: DisplayMode = DisplayMode.WORKING,
    state=None,
) -> None:
    if not isinstance(data, dict):
        console.print_text(f"  Result: {data}", style=Colors.SUCCESS)
        return

    # Non-injection completions from intent_clarification (TUI mode).
    # The LLM response was already streamed as tokens — skip the result panel.
    payload = data.get("data") if isinstance(data.get("data"), dict) else {}
    if payload.get("confirmed_intent") in ("chat", "recover"):
        return

    status = data.get("status", "unknown")
    task_state = data.get("task_state", "")

    # Determine visual style based on status
    if status == "success" or task_state in ("injected", "recovered"):
        icon = Icons.SUCCESS
        border_color = Borders.RESULT_SUCCESS
        title_text = "INJECTION SUCCESS" if task_state == "injected" else "TASK SUCCESS"
        title_style = Colors.SUCCESS
    elif status == "partial" or task_state == "partial_recovered":
        icon = Icons.WARNING
        border_color = Borders.RESULT_PARTIAL
        title_text = "PARTIAL SUCCESS"
        title_style = Colors.WARNING
    else:
        icon = Icons.FAIL
        border_color = Borders.RESULT_FAIL
        title_text = "INJECTION FAILED" if task_state else "TASK FAILED"
        title_style = Colors.ERROR

    # Build title
    title = Text()
    locator = _latest_experiment_locator(state)
    if locator:
        title.append(f" [{locator}]", style=f"bold {title_style}")
    title.append(f" {icon} ", style=f"bold {title_style}")
    title.append(title_text, style=f"bold {title_style}")

    # Build body with structured metadata.
    #
    # PR-C2: the metadata block was a hand-padded ``label:`` + ``ljust(16)``
    # loop. ``ljust`` aligns by Python char count, but in a CJK locale a
    # Han glyph occupies 2 terminal cells, so a Chinese fault_type used to
    # shove the value into the next column. We now render the metadata via
    # ``make_field_table`` (Rich's column-width logic is cell-aware) and
    # keep the rest of the panel as Text via a Group composition.
    meta_fields: list[tuple[str, object]] = []
    if task_id:
        meta_fields.append(("Task ID", task_id))
    if data.get("fault_type"):
        meta_fields.append(("Fault Type", data["fault_type"]))
    if data.get("blade_uid"):
        meta_fields.append(("Blade UID", data["blade_uid"]))
    if task_state:
        meta_fields.append(("State", task_state))

    meta_renderable = (
        make_field_table(
            meta_fields,
            label_min_width=10,
            value_min_width=30,
            boxed=False,
            indent=2,
        )
        if meta_fields
        else None
    )

    body = Text()

    # Replan history — agent self-improvement timeline.
    # _read_diagnostic so the renderer works for both flat-dict tests and
    # production envelopes where the field lives at envelope["data"][key].
    replan_history = _read_diagnostic(data, "replan_history", []) or []
    if replan_history:
        _render_replan_history(body, replan_history, status)

    # Verification section
    verification = data.get("verification")
    if verification:
        body.append("  \u2500\u2500 Verification ", style=Colors.DIM)
        body.append("\u2500" * 30 + "\n", style=Colors.DIM)
        if isinstance(verification, str):
            for line in verification.strip().splitlines():
                line = line.strip()
                if line:
                    # Color code verification lines
                    if any(kw in line.lower() for kw in ("pass", "ok", "success", "\u2713")):
                        body.append(f"  {Icons.SUCCESS} ", style=Colors.SUCCESS)
                    elif any(kw in line.lower() for kw in ("fail", "error", "\u2717")):
                        body.append(f"  {Icons.FAIL} ", style=Colors.ERROR)
                    else:
                        body.append("  \u2022 ")
                    body.append(f"{line}\n")
        elif isinstance(verification, dict):
            for k, v in verification.items():
                body.append(f"  {Icons.SUCCESS} ", style=Colors.SUCCESS)
                body.append(f"{k}: {v}\n")

    # Side effects — chaos-eng-specific value signal.
    # ``container_restarts`` means the fault triggered a real pod restart
    # (evidence got destroyed but the experiment was effective). The data
    # has always been recorded by the verifier; the user just never saw it.
    side_effects = _read_diagnostic(data, "side_effects", {}) or {}
    if isinstance(side_effects, dict):
        _render_side_effects(body, side_effects)

    # PR-D5 physical timeline. Renders working/dense; calm is silent.
    # Data is pre-existing in the result envelope (created_at /
    # injection_start_time / finished_at), we only re-shape it.
    _render_physical_timeline(body, data, display_mode)

    # Error details for failed results.
    # Prefer the structured failure_detail (category + context + hint) over
    # the raw merged_error so failed runs show *why* on one line and the
    # LLM diagnosis on a second, instead of a single ungrokable blob.
    if status not in ("success",):
        cause, hint = _extract_failure_cause_hint(data)
        if cause:
            body.append("  Cause: ", style=f"bold {Colors.ERROR}")
            body.append(f"{cause}\n", style=Colors.ERROR)
            if hint:
                body.append("  Hint:  ", style=f"bold {Colors.MUTED}")
                body.append(f"{hint}\n", style=Colors.MUTED)
        else:
            error_msg = _read_diagnostic(data, "error", None) or data.get("message")
            if error_msg:
                body.append(f"  Error: {error_msg}\n", style=Colors.ERROR)

    # Recovery info
    recovery = data.get("recovery")
    if recovery:
        body.append("\n")
        body.append("  \u2500\u2500 Recovery ", style=Colors.DIM)
        body.append("\u2500" * 33 + "\n", style=Colors.DIM)
        if isinstance(recovery, str):
            body.append(f"  {Icons.SUCCESS} ", style=Colors.SUCCESS)
            body.append(f"{recovery}\n")
        elif isinstance(recovery, dict):
            for k, v in recovery.items():
                body.append(f"  {k}: {v}\n")

    panel_body = (
        Group(meta_renderable, body) if meta_renderable is not None else body
    )
    # Vertical breath: 1 blank line before result card
    console.print("")
    console.print(
        Panel(
            panel_body,
            title=title,
            border_style=border_color,
            padding=Spacing.RESULT_PADDING,
        )
    )
    # Bell notification on completion
    console.bell()
